### PR TITLE
Atomic-free JNI work

### PR DESCRIPTION
**Pull request by gacholio**
- add some stronger barriers to exclusive
- refactor exclusive to check inNative first

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

File: runtime/vm/VMAccess.cpp [hunk 1]
```diff
@@ -261,18 +270,15 @@ acquireExclusiveVMAccess(J9VMThread * vmThread)
...
 #endif /* !J9VM_INTERP_TWO_PASS_EXCLUSIVE */
-					} else {
+				} else if (currentThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS) {
 #if defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
-						VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
+					VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
 #endif /* J9VM_INTERP_TWO_PASS_EXCLUSIVE */
-						responsesExpected++;
-					}
+					responsesExpected++;
```
Comment (DanHeidinga) [inline runtime/vm/VMAccess.cpp]: Can this and the `#else` branch be merged into 1 copy?
```
#if defined(J9VM_INTERP_ATOMIC_FREE_JNI)
				if (currentThread->inNative) {
#if !defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
					VM_VMAccess::setPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
#endif /* !J9VM_INTERP_TWO_PASS_EXCLUSIVE */
				} else 
#endif /* J9VM_INTERP_ATOMIC_FREE_JNI */
				if (currentThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS) {
#if defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
					VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
#endif /* J9VM_INTERP_TWO_PASS_EXCLUSIVE */
					responsesExpected++;
				}
```

File: runtime/vm/VMAccess.cpp [hunk 2]
```diff
@@ -754,18 +769,15 @@ acquireExclusiveVMAccessFromExternalThread(J9JavaVM * vm)
...
 #endif /* !J9VM_INTERP_TWO_PASS_EXCLUSIVE */
-			} else {
+		} else if (currentThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS) {
 #if defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
-				VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
+			VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
 #endif /* J9VM_INTERP_TWO_PASS_EXCLUSIVE */
-				vmResponsesExpected++;
-			}
+			vmResponsesExpected++;
```
Comment (DanHeidinga) [inline runtime/vm/VMAccess.cpp]: same question about code duplication

File: runtime/vm/VMAccess.cpp (hunk 1 above)
Comment (gacholio) [reply runtime/vm/VMAccess.cpp]: I'm intentionally keeping the old and new paths separate for now.  Once the new feature has been proven, the old code will be removed.

File: runtime/vm/VMAccess.cpp (hunk 2 above)
Comment (gacholio) [reply runtime/vm/VMAccess.cpp]: Same comment as above.

**Comment by gacholio** [comment]
I added the stronger barriers for safety on platforms I'm not completely sure about.  This is the slow path, so there should be no performance impact.

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk8

**Pull request merged**